### PR TITLE
bpo-37085: Expose SocketCAN bcm_msg_head flags

### PR DESCRIPTION
**Pull request by karlding**
Expose the CAN_BCM SocketCAN constants used in the bcm_msg_head struct
flags (provided by <linux/can/bcm.h>) under the socket library.

This adds the following constants with a CAN_BCM prefix:

  * SETTIMER
  * STARTTIMER
  * TX_COUNTEVT
  * TX_ANNOUNCE
  * TX_CP_CAN_ID
  * RX_FILTER_ID
  * RX_CHECK_DLC
  * RX_NO_AUTOTIMER
  * RX_ANNOUNCE_RESUME
  * TX_RESET_MULTI_IDX
  * RX_RTR_FRAME
  * CAN_FD_FRAME

https://bugs.python.org/issue37085

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
tests?

**Comment by karlding** [comment]
Could you clarify what you mean, @auvipy?

These are just binding the C `#define`s, so do you want me to test that these values are some known value? If so, wouldn't that just be testing the Linux implementation?

Thanks!

File: Modules/socketmodule.c
Comment (taleinat) [inline Modules/socketmodule.c]: Why these separate `#ifdef` checks for each constant? Are they not either all available together, or not? More specifically, isn't the `#ifdef HAVE_LINUX_CAN_BCM_H` check enough to ensure their availability?

**Review by taleinat — CHANGES_REQUESTED** [review]
This certainly needs tests, at least checking availability as done for the other `socket.CAN_BCM_*` constants in `test_socket.BasicCANTest.testBCMConstants()`.

File: Modules/socketmodule.c
Comment (karlding) [reply Modules/socketmodule.c]: This is not necessarily true for all the constants, namely the backwards-compatible introduction of the `CAN_FD_FRAME` flag, which was added in `6f3b911d5f29b98752e5da86a295210c0c4f4e14` as part of the the `4.8.x` kernel series.

The other constants were always present since the SocketCAN drivers were mainlined in `2.6.25`.

Should I just have the `#ifdef` on `CAN_FD_FRAME` then? I had added the other ones for consistency.

Comment (taleinat) [reply Modules/socketmodule.c]: > Should I just have the #ifdef on CAN_FD_FRAME then? I had added the other ones for consistency.

IMO that would be better, yes. And perhaps add a note about the availability of `CAN_FD_FRAME` in [the docs](https://docs.python.org/3/library/socket.html#socket.CAN_BCM).

File: Lib/test/test_socket.py
Comment (karlding) [inline Lib/test/test_socket.py]: Is it better to parse the kernel version from `platform.release` as a way of checking whether this exists?

Comment (taleinat) [reply Lib/test/test_socket.py]: Yes, using the `test.support.requires_linux_version` decorator.

Comment (taleinat) [reply Lib/test/test_socket.py]: And don't forget to also use the same decorator as above, i.e.:
```python
    @unittest.skipUnless(hasattr(socket, "CAN_BCM"),
                         'socket.CAN_BCM required for this test.')
    @requires_linux_version(4, 8)
    def testBCMConstantsSupportsFD(self):
        # ...
```

Comment (taleinat) [reply Lib/test/test_socket.py]: Also the method name for the separate test is unclear to me ("supports FD"). Either of `testBCMConstantsLinux48` or `testBCMConstantsFD` seems better.

**Comment by karlding** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Library/2019-06-18-16-29-31.bpo-37085.GeYaD6.rst
Comment (taleinat) [inline Misc/NEWS.d/next/Library/2019-06-18-16-29-31.bpo-37085.GeYaD6.rst]: Some BCM constants were already there; perhaps more precisely describe the nature of the added constants/flags.

**Comment by taleinat** [comment]
Thanks, @karlding! This is looking good, I have just a small comment on the NEWS entry.

**Comment by taleinat** [comment]
The CI tests are failing due to the `CAN_BCM_CAN_FD_FRAME` flag being missing, on Linux (specifically, Ubuntu 16.04 on both Travis and Azure Pipelines).

**Comment by taleinat** [comment]
> The CI tests are failing due to the `CAN_BCM_CAN_FD_FRAME` flag being missing, on Linux (specifically, Ubuntu 16.04 on both Travis and Azure Pipelines).

I ran the tests with this PR on an Ubuntu 18.10 installation and they passed. Perhaps get access to a host running Ubuntu 16.04 to reproduce and diagnose the issue.

**Comment by karlding** [comment]
So correct me if I'm wrong, but I believe this is because the Xenial headers included in `linux-libc-dev` for userspace (ie. `glibc`) in the repos are for an older version (`4.4.x`) of the kernel.

Looking at `/usr/include/linux/version.h` for a clean 16.04 install, we see:

```c
#define LINUX_VERSION_CODE 263335
```

which corresponds to the following (`4.4.167`):

```c
(((4) << 16) + ((4) << 8) + (167))
```

And even after updating `linux-libc-dev`, we see

```c
#define LINUX_VERSION_CODE 263349
```

which corresponds to `4.4.181`:

```c
(((4) << 16) + ((4) << 8) + (181))
```

Despite running a `4.15` kernel:

```
karl@karl-VirtualBox:/usr/include$ uname -a
Linux karl-VirtualBox 4.15.0-45-generic #48~16.04.1-Ubuntu SMP Tue Jan 29 18:03:48 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

I'm not too familiar with packaging, but if I'm reading this [Ubuntu kernel release schedule](https://wiki.ubuntu.com/Kernel/Support) page correctly, this seems to indicate that they ship `4.4.x` headers.

And according to this [AskUbuntu post](https://askubuntu.com/questions/957411/what-is-linux-libc-dev-linux-kernel-headers-for-development-for-another-ker), in Ubuntu LTS releases, it appears that they are not upgraded with the HWE kernels. So while the kernel may support it, the headers provided by the repos don't provide the necessary userspace API.

**Comment by taleinat** [comment]
@karlding, great research and write-up!

I think the existing implementation approach, exposing the flags only when they are available at the C level, is good. It's just the tests that need to be fixed.

In this case even just not testing for this flag would be fine IMO, but here I'd definitely want another core dev's opinion. So please remove it from the test, make sure tests are passing, and we'll wait for another reviewer.

@serhiy-storchaka, any chance for a quick review of this, perhaps?

**Comment by karlding** [comment]
Hmm, looking through the logs, the Azure Pipelines macOS failure doesn't seem to be related to my change. Should I just rebase and then trigger a new CI job?

The previously failing tests are now passing:

Travis:

```
0:12:32 load avg: 3.34 [407/419] test_socket passed
```

Azure Pipelines:

```
0:07:35 load avg: 5.36 [337/419] test_socket passed (30 sec 399 ms) -- running: test_concurrent_futures (2 min 38 sec), test_multiprocessing_forkserver (1 min 31 sec)
```

**Comment by taleinat** [comment]
> Hmm, looking through the logs, the Azure Pipelines macOS failure doesn't seem to be related to my change.

Indeed, that failure is unrelated.

> Should I just rebase and then trigger a new CI job?

Our Azure Pipelines CI isn't stable yet and is not required to pass, so there's no need to trigger a new run.

(Also, please avoid using rebase on an existing PR branch, it makes reading its history much more confusing.)

**Comment by taleinat** [comment]
@karlding, which current OS-s *do* support the `CAN_BCM_CAN_FD_FRAME` and provide it in their headers? Perhaps we can recognize at least one or two such OS-s and limit the test for the flag to those. Even if they're not one of those triggered here on GitHub, they're likely found in our [buildbot fleet](https://buildbot.python.org/all/#/builders).

File: Modules/socketmodule.c
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]:
```suggestion
#ifdef CAN_FD_FRAME
```

File: Doc/library/socket.rst
Comment (serhiy-storchaka) [inline Doc/library/socket.rst]:
```suggestion
      The :data:`CAN_BCM_CAN_FD_FRAME` flag is only available on Linux >= 4.8.
```

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM, just fix some spacing.

**Comment by serhiy-storchaka** [comment]
AFAIK, the only OS besides Linux which supports CAN_BCM is NetBSD starting from version 8. But our NetBSD buildbots were broken and removed a long time ago.

**Comment by taleinat** [comment]
> AFAIK, the only OS besides Linux which supports CAN_BCM is NetBSD starting from version 8. But our NetBSD buildbots were broken and removed a long time ago.

I was thinking of other Linux distros. Actually, I myself checked that it is available on Ubuntu 18.10 (see previous comment). I suggest to check on 18.04 (since it's LTS), and if found there, enable the removed tests just on Ubuntu >= 18.04.

**Comment by karlding** [comment]
> I suggest to check on 18.04 (since it's LTS), and if found there, enable the removed test just on Ubuntu >= 18.04.

I checked on Ubuntu 18.04, and after installing `linux-libc-dev` I see that the constant is present.

```
karl@karl-VirtualBox:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic
karl@karl-VirtualBox:~$ tail -n5 /usr/include/linux/can/bcm.h
#define TX_RESET_MULTI_IDX  0x0200
#define RX_RTR_FRAME        0x0400
#define CAN_FD_FRAME        0x0800

#endif /* !_UAPI_CAN_BCM_H */
```

With that being said, is there a non-deprecated way to check for Ubuntu versions (ie. besides `platform.linux_distribution`)? Or could you point me to an example or any documentation where you're only running the test on a particular builder type?

**Comment by karlding** [comment]
The constant should also be available on Debian Stretch and beyond, since `linux-libc-dev` seems to provide `4.9.x` headers in their repos.

**Comment by taleinat** [comment]
> With that being said, is there a non-deprecated way to check for Ubuntu versions (ie. besides `platform.linux_distribution`)?

That's a good question, I hadn't realized it had been deprecated without any replacement.

> Or could you point me to an example or any documentation where you're only running the test on a particular builder type?

No, I can't find an example of this in our codebase.

**Review by serhiy-storchaka — APPROVED** [review]
I think it can be merged without test for CAN_BCM_CAN_FD_FRAME.

**Comment by karlding** [comment]
Also, should this be backported, since the docs (incorrectly) indicate that all the BCM constants are available on older versions?

**Comment by serhiy-storchaka** [comment]
It may be backported to 3.8 (ask @ambv for this), but not to 3.7.

**Comment by taleinat** [comment]
Shouldn't we add a note about this in "What's New"?

**Pull request merged**
